HW1/P4/P4_bfs.py

```diff
@@ -0,0 +1,109 @@
+###################################################
+### Problem 4 - Graph processing in Spark [25%] ###
+### P4_bfc.py									###
+### Patrick Day 								###
+### CS 205 HW1                                  ###
+### Oct 4th, 2015								###
+###################################################
+
+########################
+### Import Functions ###
+########################
+import pyspark
+import numpy as np
+import csv
+from operator import add
+import itertools
+
+import os
+sc = pyspark.SparkContext()
+sc.setLogLevel('WARN')
+
+########################
+### Define Functions ###
+########################
+def copartitioned(RDD1, RDD2):
+	"check if two RDDs are copartitioned"
+	return RDD1.partitioner == RDD2.partitioner
+
+def build_graph(lines):
+	### Build the hero graph ###
+	hero_rdd = sc.parallelize(lines).partitionBy(10)
+	comic_rdd = hero_rdd.map(lambda x: (x[1], x[0]))#, preservesPartitioning=True)
+	
+	# Group all comics together by key
+	comic_list = comic_rdd.groupByKey().mapValues(list).collect()
+	comic_flat_rdd = sc.parallelize(comic_list).partitionBy(10)
+	
+	# Create hero pairs
+	hero_pairs = comic_flat_rdd.flatMap(lambda x: list(itertools.combinations(x[1], 2)))
+	hero_pair_rev = hero_pairs.map(lambda x: (x[1], x[0]), preservesPartitioning=True)
+	
+	# Bring all pairs back together and group
+	all_heros = hero_pairs.union(hero_pair_rev)
+	hero_combine = all_heros.groupByKey().mapValues(list)
+
+	# Preprocess the graph according to (k=Hero, (v0=depth, v1=seen/processed/not_see))
+	ref_graph = hero_combine.map(lambda x: (x[0], list(set(x[1]))), preservesPartitioning=True)
+	graph_init = ref_graph.map(lambda x: (x[0], (x[1], 100, 0))).partitionBy(10).cache()
+
+	return graph_init
+
+def ss_bfs(graph_init, hero, partition=10):
+	
+	### Initialize first iteration ###
+	
+	# Create Hero RDD and set parent to processed
+	hero_values = graph_init.filter(lambda x: x[0] == hero)
+	processed_parent = hero_values.map(lambda x: (x[0], (0, 2)), preservesPartitioning=True)
+	parent = hero_values.map(lambda x: (x[0][0], (x[1][0], 0, 1)), preservesPartitioning=True)
+	
+	# Update parent, find neighbors, and put together
+	nei_init_rdd = parent.flatMap(lambda x: x[1][0]).map(lambda x: (x,(1,1)), preservesPartitioning=True).subtractByKey(processed_parent).partitionBy(partition)
+	hero_graph = nei_init_rdd.union(processed_parent).partitionBy(partition).cache()
+	
+	# Initialize accumlators and depth after preprocessing
+	accum_par = sc.accumulator(0)
+	accum_nei = sc.accumulator(0)
```

Comment:
Seems like your accumulators just act as variables -- there are other ways you can use accumulators to serve some useful functions as well